### PR TITLE
Fix overflow when reading negative s8

### PR DESCRIPTION
**Pull request by cherue**
When doing bitwise operations in JS the values are always converted to a ~~signed~~ (edit: not always signed, see [here](https://tc39.es/ecma262/#sec-numberbitwiseop)) 32-bit integer and the result is also always a signed 32-bit integer.

The calculation that converts two `u4`s to one `s8` assumes only positive inputs, this change makes sure of that.

I discovered this when writing a test for JS 53-bit integer overflows, but that test and the error on overflow aren't part of this.

### CODE REVIEW

**Comment by GreyCat** [comment]
@cherue, can you clarify which test shows the problem in JS right now?

**Comment by cherue** [comment]
Right now no test shows the problem.

I am writing a test called `js_overflow` that will test that the JS runtime throws an error when reading numbers smaller than `Number.MIN_SAFE_INTEGER` or bigger than `Number.MAX_SAFE_INTEGER` (this is not implemented yet).

The ksy for this test currently looks like this:
```yaml
meta:
  id: js_overflow
seq:
  - id: signed_negative_be
    type: s8be
  - id: signed_negative_le
    type: s8le
  - id: signed_positive_be
    type: s8be
  - id: signed_positive_le
    type: s8le
  - id: unsigned_be
    type: u8be
  - id: unsigned_le
    type: u8le
instances:
  overflow_signed_negative_be:
    pos: 48
    type: s8be
  overflow_signed_negative_le:
    pos: 56
    type: s8le
  overflow_signed_positive_be:
    pos: 64
    type: s8be
  overflow_signed_positive_le:
    pos: 72
    type: s8le
  overflow_unsigned_be:
    pos: 80
    type: u8be
  overflow_unsigned_le:
    pos: 88
    type: u8le
```
While writing this test I noticed that `signed_negative_be` and `signed_negative_le` were wrong. See 
[js_overflow.bin.zip](https://github.com/kaitai-io/kaitai_struct_javascript_runtime/files/4336425/js_overflow.bin.zip) for the binary file.

Also I misnamed my branch, it is the low bytes that overflow. The high bytes can't ever overflow because the sign bit is always set. I'll fix that and then write up a better explanation of the problem.

**Comment by cherue** [comment]
Let's look at `Number.MIN_SAFE_INTEGER` or `-9007199254740991` or `0x ff e0 00 00 00 00 00 01` (big-endian) to demonstrate the problem.

First, the `s8` is read as two `u4`s:
```js
var high = this.readU4be(); // 0x ff e0 00 00
var low  = this.readU4be(); // 0x 00 00 00 01
```

Then `high` and `low` are XORed with `0x ff ff ff ff`:
```js
high = high ^ 0xffffffff; // 0x 00 1f ff ff
low  = low  ^ 0xffffffff; // 0x ff ff ff fe
```

And finally they are combined to create the `s8`:
```js
return -(0x100000000 * high + low) - 1; 
```
The expected result is `-9007199254740991` but currently the result is `-9007194959773695` (WebIDE has an outdated runtime so it's still treated as a positive number).

This happens because the result of XOR is [defined](https://tc39.es/ecma262/#sec-numberbitwiseop) to be a signed 32-bit integer. Which means the final step, which should be:
```js
return -(0x100000000 * 2097151 + 4294967294) - 1;
// -9007199254740991
```
currently is:
```js
return -(0x100000000 * 2097151 + -2) - 1;
// -9007194959773695
```

**Comment by cherue** [comment]
@GreyCat do you want a test for this first?

How about an `integer_random` test? It would be a binary file filled with random data and then instances with all integer types  and  `pos: 0` and `repeat: eos`. That should catch most edge cases. The KST and the tests would be huge though.

File: KaitaiStream.js [hunk 1]
```diff
@@ -802,6 +790,18 @@ KaitaiStream.createStringFromArray = function(array) {
   return chunks.join("");
 };
 
+KaitaiStream.twoU4sToS8 = function(high, low) {
+  if ((high & 0x80000000) != 0) {
```
Comment (generalmimon) [inline KaitaiStream.js:794]: Always use strict equality operator `===` (or `!==` negated) in JavaScript unless you have a very good reason to use _loose equality_ op with type juggling (`==`/`!=`).

```suggestion
  if ((high & 0x80000000) !== 0) {
```

The loose equality op `==` knows several ways how to surprise you: at random `"" == false`, `"" == 0`, `[] == false` (whereas [`![] === false`](https://github.com/denysdovhan/wtfjs#true-is-not-equal--but-not-equal--too) actually) , `[] == ''`, `[] == 0`, `[0] == false`, `[1] == true`, `[null, null] == ","` etc. Thus it's _rarely_ a good idea to use the loose equality.

It doesn't matter whether you think you know the type of the operands, because it's JavaScript. Unless you are doing `typeof X === ...` everywhere, you don't know the types.

File: KaitaiStream.js [hunk 2]
```diff
@@ -802,6 +790,18 @@ KaitaiStream.createStringFromArray = function(array) {
...
+    // negative number
+    high = high ^ 0xffffffff;
+    low = low ^ 0xffffffff;
+    low = low < 0 ? 2**32 + low : low;
```
Comment (generalmimon) [inline KaitaiStream.js:798]: Please don't use exponentiation operator in JavaScript, because this would noticeably impair the KS runtime compatibility with JavaScript environments (compare compatibility table on MDN of [exponentiation `**`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Exponentiation#Browser_compatibility) and [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array#Browser_compatibility), which is probably the latest JS feature the runtime uses).

&nbsp; | Chrome | Edge | Firefox | Internet Explorer | Opera | Safari | Android webview | Chrome for Android | Firefox for Android | Opera for Android | Safari on iOS | Samsung Internet | Node.js
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array#Browser_compatibility) | 7 | 12 | 4 | 10 | 11.6 | 5.1 | 4 | 18 | 4 | 12 | 4.2 | 1.0 | 0.10
[Exponentiation (`**`)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Exponentiation#Browser_compatibility) | 52 | 14 | 52 | **No** | 39 | 10.1 | 51 | 52 | 52 | 41 | 10.3 | 6.0 | 7.0.0

Moreover, environments that don't know about exponential operator `**` treat its usage anywhere in the code as a syntax error, regardless of whether the code using it would run or not. So it's impossible to polyfill it in such environments, because it's a syntax thing. This is a significant difference from `Uint8Array`, which can be polyfilled.

I believe that in theory you can bring the current KS runtime up and running even in ancient browsers like IE 5 🙂, if you link a few polyfills.

I suggest a cleaner (and probably also faster) way how to convert the 32-bit signed integer to an unsigned one - using the [unsigned right shift](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Unsigned_right_shift) operator. From [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Unsigned_right_shift):

> Unlike the other bitwise operators, zero-fill right shift returns an unsigned 32-bit integer.

So it's enough to do `>>> 0`, which is guaranteed to yield a 32-bit unsigned integer. You don't even have to check if the original number is positive or negative.

```suggestion
    low >>>= 0; // convert to unsigned 32-bit integer
```

**Review by generalmimon — CHANGES_REQUESTED** [review]
